### PR TITLE
refactor: replace bash script with python

### LINKED ISSUE #48: Replace bash with Python
I don't think I am alone in that complicated bash scripts are hard to read, so I suggest replacing the bash script for finding changed functions with Python. Note GitHub Actions comes with Python 3.10 preinstalled, so there is no performance hit on this. 

I have already written such a script for a project [here](https://github.com/cognitedata/tech-demo-powerops/blob/main/build_scripts/changed_functions.py) and I am happy to add it here. 

@haakonvt Are you controlling this repo? wdyt?

### PR DESCRIPTION
**Pull request by doctrino**
closes #48 

For arguments for why see issue. 

This script is battle tested in [https://github.com/cognitedata/tech-demo-powerops]( https://github.com/cognitedata/tech-demo-powerops)

### CODE REVIEW

**Review by luka-mikec — APPROVED** [review]

File: build_scripts/changed_functions.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        nargs=1,
+        help="The folders to monitor for deployment to cognite functions",
+    )
+    parser.add_argument("deploy_all", type=str, nargs="?", help="The folders to deploy to all ", default=None)
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    print(f"Got funtion folders: {args.folders!r}")  # noqa
```
Comment (haakonvt) [inline build_scripts/changed_functions.py:23]:
```suggestion
    print(f"Got function folders: {args.folders!r}")  # noqa
```

File: build_scripts/changed_functions.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    print(f"And deploy all folders: {args.deploy_all!r}")  # noqa
+    function_folders = {f.strip() for f in args.folders[0].split(",")}
+    deploy_all_folders = (args.deploy_all or {}) and {f.strip() for f in args.deploy_all.split(",")}
+
+    process = subprocess.Popen("git diff --name-only HEAD^ HEAD".split(" "), stdout=subprocess.PIPE)  # nosec
```
Comment (haakonvt) [inline build_scripts/changed_functions.py:28]: Doesn't `HEAD^ HEAD` only check vs the previous commit? 🤔 

Also `split` splits on whitespace by default 🪄 
```suggestion
    process = subprocess.Popen("git diff --name-only master".split(), stdout=subprocess.PIPE)  # nosec
```

**Review by haakonvt — CHANGES_REQUESTED** [review]

File: build_scripts/changed_functions.py (hunk 2 above)
Comment (doctrino) [reply build_scripts/changed_functions.py]: Yes it is correct that `HEAD^ HEAD` only checks against the previous commit, thus you are required to use squash commits. This is the same behavior as the previous bash script. Unfortunately, I don't think your suggestion will work as this workflow is expected to run on the default branch, master in this case, and thus there are no diffs between master and master and nothing will be triggered.

Comment (haakonvt) [reply build_scripts/changed_functions.py]: Ah, of course 🤦‍♂️ How do we make sure people squash? Seems unsafe :( I also see that the previous version did it like this. I think we need to document this loud and clear in the how-to-section; something along the lines of _the way changes are picked up is by diffing vs the previous commit, thus you risk functions not deploying unless you always squash to master. You can control this behaviour in the Github repo settings <image>_

Comment (haakonvt) [reply build_scripts/changed_functions.py]: Preferrably with a screenshot attached

Comment (doctrino) [reply build_scripts/changed_functions.py]: Updated documentation with the preferrable screenshot :)

**Review by haakonvt — APPROVED** [review]

**Pull request merged**